src/main/java/org/openrewrite/github/SetupJavaUpgradeJavaVersion.java

```diff
@@ -0,0 +1,73 @@
+package org.openrewrite.github;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+import org.openrewrite.*;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.yaml.JsonPathMatcher;
+import org.openrewrite.yaml.YamlVisitor;
+import org.openrewrite.yaml.tree.Yaml;
+
+@Value
+@EqualsAndHashCode(callSuper = true)
+public class SetupJavaUpgradeJavaVersion extends Recipe {
+
+    @Option(displayName = "Minimum major Java version (defaults to 17)",
+            example = "17",
+            required = false)
+    @Nullable
+    Integer minimumJavaMajorVersion;
+
+    @Override
+    public String getDisplayName() {
+        return "Upgrade `actions/setup-java` `java-version`";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Update the Java version used by `actions/setup-java` if it is below the expected version number.";
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getSingleSourceApplicableTest() {
+        return new HasSourcePath<>(".github/workflows/*.yml");
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new UpgradeJavaVersionVisitor(
+                minimumJavaMajorVersion == null ? 17 : minimumJavaMajorVersion
+        );
+    }
+
+    private static class UpgradeJavaVersionVisitor extends YamlVisitor<ExecutionContext> {
+        private static final JsonPathMatcher javaVersion = new JsonPathMatcher("..steps[?(@.uses =~ 'actions/setup-java@v*.*')].with.java-version");
+
+        private final int minimumJavaMajorVersion;
+
+        public UpgradeJavaVersionVisitor(int minimumJavaMajorVersion) {
+            this.minimumJavaMajorVersion = minimumJavaMajorVersion;
+        }
```

Comment:
Since we're already using lombok, might as well use it here, too. Could slap an `@AllArgsConstructor` or `@Data` on this class and delete the constructor.